ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -759,12 +750,14 @@ delayedAction a = do
 -- | Restart the current 'ShakeSession' with the given system actions.
 --   Any actions running in the current session will be aborted,
 --   but actions added via 'shakeEnqueue' will be requeued.
-shakeRestart :: Recorder (WithPriority Log) -> IdeState -> VFSModified -> String -> [DelayedAction ()] -> IO ()
-shakeRestart recorder IdeState{..} vfs reason acts =
+shakeRestart :: Recorder (WithPriority Log) -> IdeState -> VFSModified -> String -> [DelayedAction ()] -> IO [Key] -> IO ()
+shakeRestart recorder IdeState{..} vfs reason acts ioActionBetweenShakeSession =
     withMVar'
         shakeSession
         (\runner -> do
               (stopTime,()) <- duration $ logErrorAfter 10 $ cancelShakeSession runner
+              keys <- ioActionBetweenShakeSession
+              atomically $ modifyTVar' (dirtyKeys shakeExtras) $ \x -> foldl' (flip insertKeySet) x keys
               res <- shakeDatabaseProfile shakeDb
               backlog <- readTVarIO $ dirtyKeys shakeExtras
```

Comment:
Do we even need the `dirtyKeys` TVar now? I believe we now only set it just above and then we immediately read it!

Reply:
yes, I am thinking about removing it too, but our garbage collector depend on it.

Reply:
So we have to keep it this way for now. Maybe until we find a better way to handle the garbage collector, but this should be in another PR.

Reply:
It looks to me like the garbage collector calls `getDirtySet` which gets the dirty keys from `hls-graph`?

Reply:
Yes, but it also mark the key as dirty at `foldM (removeDirtyKey dirtyKeys state) (0,[]) agedKeys` accompany to the rule cache removal.

Reply:
This is odd since we might have clean it in the rule cahce, 
the a part of the rule cahce would still stays in the hls-graph database.
We might want to improve this in the future.